### ISSUE TITLE
Discrepancy between PHP and hhvm with generators

### ISSUE THREAD

**Issue by robations** [ReactiveX/RxPHP#23]
See this code sample:

https://gist.github.com/anonymous/2611cbe9564fd820b9b2

To summarise the output, PHP generates expected output returning numbers 0–12 for both the subscribe and foreach.

hhvm outputs 1–12 for the subscribe and 0–12 for the foreach.

```
$ php -v
PHP 5.6.17 (cli) (built: Jan  8 2016 10:27:48) 
Copyright (c) 1997-2015 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2015 Zend Technologies
    with Xdebug v2.3.3, Copyright (c) 2002-2015, by Derick Rethans

$ hhvm --version
HipHop VM 3.11.0 (rel)
Compiler: 1454683423_N
Repo schema: abdeb80638110f1c5b97727498d253ab662076cf
```

(Tested against rxphp 1.1.0)

**Comment by robations** [comment]
Seems like undo-ing [this fix](https://github.com/ReactiveX/RxPHP/commit/5df42cac057743ca755fa3c46dec68bfe421eb67) resolves the issue for me. I wonder if there is a bug that's been fixed in hhvm 3.11?

Edit: I tried looking through the facebook/hhvm issue log, but the landscape is confusing.

**Comment by davidwdan** [comment]
That's very interesting.  The change must be between v3.10 and 3.11.  I get the correct results with 3.10.1

**Comment by davidwdan** [comment]
Looks like this is a related issue:  https://github.com/facebook/hhvm/issues/6807

**Comment by robations** [comment]
I think that's the issue that was confusing me. Definitely related, but jeskew seems to be saying that it is _3.11_ that is inconsistent with PHP, whereas here it seems to be 3.10 that was at odds. So Travis-CI should be on 3.11?

Edit: saying 'seems' too much

**Comment by robations** [comment]
Confirmed this behaviour with hhvm 3.11.0 on OS X and Ubuntu.

I started out writing code here but decided easier just to create a pull request: https://github.com/ReactiveX/RxPHP/pull/30

I've not read any submission guidelines so the PR is provided more as a discussion point than anything else.

I reckon it could be a problem running the version comparison inside a loop so I've taken the comparison outside the anonymous function. Not super keen about effectively duplicating the anon fn but it could be easier to understand than the if statements. And I just thought of another way to refactor this, but my egg-timer just ran out...
